numpy/core/src/multiarray/convert.c

```diff
@@ -613,11 +613,15 @@ PyArray_View(PyArrayObject *self, PyArray_Descr *type, PyTypeObject *pytype)
         subtype = Py_TYPE(self);
     }
 
-    if (type != NULL && (PyArray_FLAGS(self) & NPY_ARRAY_WARN_ON_WRITE)) {
+    flags = PyArray_FLAGS(self);
+    dtype = PyArray_DESCR(self);
+
+    if (type != NULL && !PyArray_EquivTypes(dtype, type) &&
```

Comment:
The addition of `PyArrayEquivTypes` here, (the only real change), is so that code in `records.py` like `arr.view((np.record, arr.dtype))` doesn't complain. 

That kind of view doesn't change the memory layout so is save to keep in 1.16.